File: java/src/priorityQueue/HeapSort.java
package priorityQueue;

import java.util.Arrays;
import java.util.*;

public class HeapSort {
    public void heapSortNotInPlace(int[] A, int size) {
        PirorityQeuueIn pq = new PirorityQeuueIn();
        for (int i : A)
            pq.insert(i);
        for (int i = size - 1; i >= 0 ; i--) {
            A[i] = pq.extractMax();
        }
    }

    class PirorityQeuueIn {
        private int[] H = new int[10000];
        private int size = -1;
        private int maxSize = 9999;

        private int parent(int idx) {
            return (int) Math.floor((idx - 1) / 2);
        }

        private int leftChild(int idx) {
            return idx * 2 + 1;
        }

        private int rightChild(int idx) {
            return idx * 2 + 2;
        }

        private void swap(int i, int j) {
            int tmp = this.H[i];
            this.H[i] = this.H[j];
            this.H[j] = tmp;
        }

        private void shiftUp(int idx) {
            while (idx > 0 && this.H[idx] > this.H[this.parent(idx)]) {
                int p = this.parent(idx);
                this.swap(idx, p);
                idx = p;
            }
        }

        private void shiftDown(int idx) {
            int maxIdx = idx;

            if (idx <= this.size) {
                int l = this.leftChild(idx);
                if (l <= this.size && this.H[maxIdx] < this.H[l])
                    maxIdx = l;

                int r = this.rightChild(idx);
                if (r <= this.size && this.H[maxIdx] < this.H[r])
                    maxIdx = r;

                if (idx != maxIdx) {
                    this.swap(idx, maxIdx);
                    shiftDown(maxIdx);
                }
            }
        }

        public void insert(int p) {
            if (this.size == this.maxSize)
                throw new IndexOutOfBoundsException();
            this.H[++this.size] = p;
            this.shiftUp(this.size);
        }

        public int extractMax() {
            if (this.size == -1)
                throw new IndexOutOfBoundsException();
            int result = this.H[0];
            this.H[0] = this.H[this.size--];
            this.shiftDown(0);

            return result;
        }
    }

    public void heapSortInPlace(int[] A) {
        int size = A.length;
        this.buildHeap(A, size);
        while (size > 0) {
            this.swap(A,0, --size);
            this.shiftDown(A, 0, size);
        }
    }

    private void buildHeap(int[] A, int size) {
        int st = (int) Math.floor(size / 2);
        for (int i = st - 1; i >= 0; i--)
            this.shiftDown(A, i, size);
    }

    private void shiftDown(int[] A, int idx, int size) {
        int maxIdx = idx;

        if (idx < size) {
            int l = this.leftChild(idx);
            if (l < size && A[maxIdx] < A[l])
                maxIdx = l;

            int r = this.rightChild(idx);
            if (r < size && A[maxIdx] < A[r])
                maxIdx = r;

            if (idx != maxIdx) {
                swap(A, idx, maxIdx);
                shiftDown(A, maxIdx, size);
            }
        }
    }

    private int leftChild(int idx) {
        return idx * 2 + 1;
    }

    private int rightChild(int idx) {
        return idx * 2 + 2;
    }

    private void swap(int[] A, int i, int j) {
        int tmp = A[i];
        A[i] = A[j];
        A[j] = tmp;
    }

    public static void main(String[] args) {
        /*int[] A1 = new int[]{22, 40, 19, 3, 7, 12, 16, 19, 32, 8};
        int[] A2 = new int[]{22, 40, 19, 3, 7, 12, 16, 19, 32, 8};

        HeapSort hs = new HeapSort();
        hs.heapSortNotInPlace(A1, A1.length);
        System.out.print("Heap sort not in place: ");
        Arrays.stream(A1).forEach((a) -> System.out.print(a + " "));
        System.out.println();

        hs.heapSortInPlace(A2);
        System.out.print("Heap sort in place: ");
        Arrays.stream(A2).forEach((a) -> System.out.print(a + " "));*/

        Scanner scanner = new Scanner(System.in);
        int size = scanner.nextInt();
        int range = scanner.nextInt();

        Random random = new Random();
        HeapSort hs = new HeapSort();

        while(true) {
            int[] A1 = new int[size];
            int[] A2 = new int[size];
            for (int i = 0; i < size; i++) {
                A1[i] = A2[i] = random.nextInt(range);
            }

            Arrays.stream(A1).forEach((a) -> System.out.print(a + " "));
            hs.heapSortNotInPlace(A1, A1.length);
            hs.heapSortInPlace(A2);

            if (Arrays.equals(A1, A2)) {
                System.out.println("- pass");
            } else {
                System.out.println("- fail");
                break;
            }
        }
    }
}

File: python/priorityQueue/heapSort.py
from random import *

# ------------------------------------------------------------------------------------heap sort not in place (using priority queue)
def heapSortNotInPlace(A):
    pq = PriorityQueue()
    for a in A:
        pq.insert(a)
    for i in range(len(A) - 1, -1, -1):
        A[i] = pq.extractMax()

class PriorityQueue:
    def __init__(self):
        self.H = [None] * 10000
        self.size = 0
        self.maxSize = 10000

    def __parent(self, i):
        return i // 2
    
    def __leftChild(self, i):
        return i * 2
    
    def __rightChild(self, i):
        return i * 2 + 1
    
    def __swap(self, i, j):
        self.H[i], self.H[j] = self.H[j], self.H[i]
    
    def __shiftUp(self, i):
        while i > 1 and self.H[i - 1] > self.H[self.__parent(i) - 1]:
            p = self.__parent(i)
            self.__swap(i - 1, p - 1)
            i = p

    def __shiftDown(self, i):
        maxIdx = i
        if i <= self.size:
            l = self.__leftChild(i)
            if l <= self.size and self.H[maxIdx - 1] < self.H[l - 1]:
                maxIdx = l

            r = self.__rightChild(i)
            if r <= self.size and self.H[maxIdx - 1] < self.H[r - 1]:
                maxIdx = r

            if i != maxIdx:
                self.__swap(i - 1, maxIdx - 1)
                self.__shiftDown(maxIdx)


    def insert(self, p):
        if self.size == self.maxSize:
            raise IndexError
        self.H[self.size] = p
        self.size += 1
        self.__shiftUp(self.size)

    def extractMax(self):
        if self.size == 0:
            raise IndexError

        result = self.H[0]
        self.H[0] = self.H[self.size - 1]
        self.size -= 1
        self.__shiftDown(1)

        return result

    def remove(self, i):
        if i > self.size or i < 1:
            raise IndexError

        self.H[i - 1] = self.H[0] + 1
        self.__shiftUp(i)
        self.extractMax()

    def changePriority(self, i, p):
        if i > self.size or i < 1:
            raise IndexError

        oldp = self.H[i - 1]
        self.H[i - 1] = p
        if p > oldp:
            self.__shiftUp(i)
        else:
            self.__shiftDown(i)



# ------------------------------------------------------------------------------------heap sort in place
def leftChild(i):
    return i * 2

def rightChild(i):
    return i * 2 + 1

def swap(A, i, j):
    A[i], A[j] = A[j], A[i]

def shiftDown(A, i, size):
    maxIdx = i

    if i <= size:
        l = leftChild(i)
        if l <= size and A[maxIdx - 1] < A[l - 1]:
            maxIdx = l

        r = rightChild(i)
        if r <= size and A[maxIdx - 1] < A[r - 1]:
            maxIdx = r

        if i != maxIdx:
            swap(A, i - 1, maxIdx - 1)
            shiftDown(A, maxIdx, size)

def buildHeap(A, n):
    for i in range(n//2, 0, -1):
        shiftDown(A, i, n)

def heapSortInPlace(A):
    size = len(A)
    buildHeap(A, size)

    while size > 1:
        swap(A, 0, size - 1)
        size -= 1
        shiftDown(A, 1, size)



def main():
    A1 = [22, 40, 19, 3, 7, 12, 16, 19, 32, 8]
    A2 = A1[0:]
    heapSortNotInPlace(A1)
    print(A1)
    
    heapSortInPlace(A2)
    print(A2)

def randomTest():
    l,r = map(int, input().split())

    while True:
        A1 = [None] * l
        A2 = [None] * l
        for i in range(0, l):
            val = randint(0, r)
            A1[i] = val
            A2[i] = val

        print(A1, A2)
        heapSortNotInPlace(A1)
        heapSortInPlace(A2)

        if A1 != A2:
            print(A1, A2)
            break
        else:
            print('pass!')

if __name__ == '__main__':
    # main()
    randomTest()
